### PR TITLE
[doc] Testing Instructions

### PR DESCRIPTION
**Pull request by ppenna**
Description
-------------

This PR addresses [#356 ](https://github.com/demikernel/demikernel/issues/356)

Summary of Changes
------------------------

- Added a new document that provides detailed instructions on how to test Demikernel
- Changed Demikernel's `README.md` file to link new testing document.
- Added missing entry to `demi_timedwait()` in the manual pages.

### CODE REVIEW

File: doc/testing.md [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+> The instructions in this file assume that you have at your disposal at least one machine with Demikernel's development
+environment set up. For more information on how to set up Demikernel's development environment, check out instructions
+in the `README.md` file.
+
+## What Are Unit Tests
+
+Unit tests exercise internal components of Demikernel, individually.
+
+Demikernel has several unit tests, and all of them spread across the source tree in `demikernel/src/`. In case you want
+to check for their implementation, then search for `mod tests` and `cfg(test)` in the code base.
```
Comment (BrianZill) [inline doc/testing.md:19]: I'd change this to:

Demikernel has several unit tests spread across the source tree in `demikernel/src/`.  To check their implementation, search for `mod tests` and `cfg(test)` in the code base.

File: doc/testing.md [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    --server-addr $SERVER_IPV4_ADDR \
+    --client-addr $CLIENT_IPV4_ADDR
+```
+
+### Running System-Level Tests with Demikernel's Build System (ie. `make`)
+
+```bash
+# Set parameters for Demikernel's TCP/UDP stack.
+export SERVER_IPV4_ADDR=10.9.1.10:12345
+export CLIENT_IPV4_ADDR=10.9.1.11:12345
```
Comment (BrianZill) [inline doc/testing.md:103]: We shouldn't be using the "12345" port number in published documentation.  According to [the IANA port number list](https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.xhtml), this is assigned to the "Italk Chat System".  We should either pick a number from the Ephemeral port range (49152-65535), or if we do want to use something from the User port range (1024-49151), at least use an unassigned one.

If you want something easy to remember in the Ephemeral port range, "56789" or "55555" would work. In the User port range, "7890" is unassigned.

This same comment applies to the "shell" instructions below, and unfortunately also the demikernel-ci.py script as well.

File: doc/testing.md [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
```
Comment (BrianZill) [inline doc/testing.md:104]: For completeness, I'll note that the IPv4 private address ranges specified in RFC 1918 are [specifically allowed to be used in documentation](https://www.rfc-editor.org/rfc/rfc5737), so we're okay using these 10/8 prefix addresses here.

There is also space specifically reserved for documentation, including three /24 prefixes for IPv4, but I don't think we need to bother with those.

**Review by BrianZill — CHANGES_REQUESTED** [review]
Okay once comments are addressed.

**Review by BrianZill — APPROVED** [review]

File: doc/testing.md (hunk 1 above)
Comment (ppenna) [reply doc/testing.md]: Done.

File: doc/testing.md (hunk 3 above)
Comment (ppenna) [reply doc/testing.md:104]: Thanks for noting this. I changed the instructions to use the `192.0.2.0/24` address range.

**Pull request merged**